### ISSUE TITLE
Progressively degrading performance while running block writer example

### ISSUE THREAD

**Issue by mrtracy** [cockroachdb/cockroach#4196]
I started up the [block writer](https://github.com/cockroachdb/examples-go/tree/master/block_writer) example this weekend in order to see how performance has improved since the last time I was intensely running it (several months ago).  What I discovered was a startling progressive performance degradation:

```
2016/02/07 03:19:37 88 dumps were executed at 88.0/second (0 total errors)
2016/02/07 03:19:38 905 dumps were executed at 905.0/second (0 total errors)
2016/02/07 03:19:39 663 dumps were executed at 662.2/second (0 total errors)
2016/02/07 03:19:40 548 dumps were executed at 548.6/second (0 total errors)
2016/02/07 03:19:41 453 dumps were executed at 451.3/second (0 total errors)
2016/02/07 03:19:42 411 dumps were executed at 412.1/second (0 total errors)
2016/02/07 03:19:43 397 dumps were executed at 396.2/second (0 total errors)
2016/02/07 03:19:44 377 dumps were executed at 377.7/second (0 total errors)
2016/02/07 03:19:45 327 dumps were executed at 326.9/second (0 total errors)
2016/02/07 03:19:46 321 dumps were executed at 320.7/second (0 total errors)
2016/02/07 03:19:47 314 dumps were executed at 314.2/second (0 total errors)
2016/02/07 03:19:48 280 dumps were executed at 279.2/second (0 total errors)
2016/02/07 03:19:49 271 dumps were executed at 272.3/second (0 total errors)
2016/02/07 03:19:50 261 dumps were executed at 261.0/second (0 total errors)
2016/02/07 03:19:51 246 dumps were executed at 244.8/second (0 total errors)
2016/02/07 03:19:52 235 dumps were executed at 236.1/second (0 total errors)
2016/02/07 03:19:53 222 dumps were executed at 221.5/second (0 total errors)
2016/02/07 03:19:54 211 dumps were executed at 211.3/second (0 total errors)
2016/02/07 03:19:55 199 dumps were executed at 199.1/second (0 total errors)
2016/02/07 03:19:56 207 dumps were executed at 207.0/second (0 total errors)
2016/02/07 03:19:57 203 dumps were executed at 203.0/second (0 total errors)
2016/02/07 03:19:58 192 dumps were executed at 192.0/second (0 total errors)
```

After the example starts running, performance begins quickly and permanently degrading - eventually it seems to level off at around 30 dumps (inserts) per second, representing a 97% reduction in throughput.  

Notably, *restarting the block writer causes this phenomena to reset*: it returns to its peak around 900 inserts/s, then degrades in the same fashion.

### Experiments
I ran this tool in a number of experimental configurations trying to narrow down the problem:
* I ran the test with six nodes, three nodes, and finally on a single node cluster.  The problem is present even on a single node, which eliminated replication traffic as the culprit.
* In order to see if the block writer itself was at fault, I reconfigured the example to run against postgres; postgres had no degradation and write load was very steady (at  ~3700 inserts/s). Thus, the block writer itself was not the problem.
* Because restarting the block writer "reset" the problem, my first theory is that there was some bug with the SQL session itself that caused it to slow down over time; thus, I modified the block writer to occasionally force-close all SQL connections.  However, the performance profile persisted even with a new connection.
* My next theory was that the performance was related to the size of the table: block writer deletes the existing block table whenever it starts up.  I disabled this, and found my first real clue; if the table is not deleted, the performance _remains degraded_ even after restarting the block writer.
* Next, I wanted to see if this problem persisted even if CockroachDB was reset.  I stopped the single cockroach node and restarted it, preserving the existing tables.  *This resets the performance degredation*.
* Finally, I tried changing the table being written to by the block writer; the performance problem resets every time a new table is selected, but returns when switching back to a table which was already degraded. The degradation appears to affect each table independently.

### Summary of the Issue
I believe I have narrowed down the scope of the issue: subsequent inserts to the same table on the same running CockroachDB node will progressively degrade in performance. The performance degradation affects each table independently, and the degradation is reset if CockroachDB is restarted.

Unfortunately, I am having quite a lot of trouble profiling CockroachDB using pprof, which is probably a separate issue; therefore, I am throwing this issue to the team for now, hoping that someone has an idea of what could be causing it.

@tschottdorf

**Comment by tschottdorf** [comment]
Let's see what `EXPLAIN (TRACE)` has to say about this on Monday.

**Comment by petermattis** [comment]
Ditto what @tschottdorf said. Sampling based profiling is not the right tool for tracking down what is going on. Tracing is. It will be interesting to compare a trace of a fast table vs a degraded table. It should very quickly highlight where the slowdown is coming from. In addition to `EXPLAIN (TRACE)`, there is also the Go execution tracer which I can show you how to use.

**Comment by tschottdorf** [comment]
We'll need to add more tracepoints, but seems that this should be an easy way of tracking it down. Note the `24ms` portion at `coordinator spawns` (which means that whatever happens before that is slow, likely something on the actual node execution).

`EXPLAIN (TRACE) INSERT INTO datablocks.blocks (block_id, writer_id, block_num, raw_bytes) VALUES (experimental_unique_int(), experimental_unique_int()::string, 12345, 'asd');`

```
+-------------------------------------------+--------------+-----+-----------+------------------------------------+--------+------+-------+--------+
|                 Timestamp                 |   Duration   | Pos | Operation |               Event                | RowIdx | Key  | Value | Output |
+-------------------------------------------+--------------+-----+-----------+------------------------------------+--------+------+-------+--------+
| 2016-02-07 19:11:41.666236726 +0000 +0000 | ""           |   0 | "sql"     | "sending batch"                    |      0 | NULL | ""    | NULL   |
| 2016-02-07 19:11:41.666244139 +0000 +0000 | "7.413µs"    |   1 | "sql"     | "meta descriptor lookup"           |      0 | NULL | ""    | NULL   |
| 2016-02-07 19:11:41.666267616 +0000 +0000 | "23.477µs"   |   2 | "sql"     | "sending RPC to [/Table/51, /Max)" |      0 | NULL | ""    | NULL   |
| 2016-02-07 19:11:41.666278514 +0000 +0000 | "10.898µs"   |   3 | "sql"     | "sending to [::]:26257"            |      0 | NULL | ""    | NULL   |
| 2016-02-07 19:11:41.66639758 +0000 +0000  | ""           |   0 | "node"    | "node 1"                           |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.666401054 +0000 +0000 | "3.474µs"    |   1 | "node"    | "read has no clock uncertainty"    |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.666416308 +0000 +0000 | "15.254µs"   |   2 | "node"    | "executing 3 requests"             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.666419376 +0000 +0000 | "3.068µs"    |   3 | "node"    | "read-write path"                  |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.666420119 +0000 +0000 | "743ns"      |   4 | "node"    | "command queue"                    |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.666458375 +0000 +0000 | "38.256µs"   |   5 | "node"    | "raft"                             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.668230464 +0000 +0000 | "1.772089ms" |   6 | "node"    | "applying batch"                   |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:11:41.691125564 +0000 +0000 | "24.84705ms" |   4 | "sql"     | "coordinator spawns"               |      0 | NULL | ""    | NULL   |
| 2016-02-07 19:11:41.691286824 +0000 +0000 | "161.26µs"   |   5 | "sql"     | "tracing completed"                |      0 | NULL | ""    | NULL   |
+-------------------------------------------+--------------+-----+-----------+------------------------------------+--------+------+-------+--------+
```

**Comment by tschottdorf** [comment]
extract from early in and later, with added tracepoints:

```
| 2016-02-07 19:20:01.519570247 +0000 +0000 | "36.241µs"   |   5 | "node"    | "raft"                             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:20:01.519922547 +0000 +0000 | "352.3µs"    |   6 | "node"    | "applying batch"                   |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:20:01.519924189 +0000 +0000 | "1.642µs"    |   7 | "node"    | "check sequence cache"             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:20:01.519940667 +0000 +0000 | "16.478µs"   |   8 | "node"    | "execution"                        |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:20:01.527763144 +0000 +0000 | "7.822477ms" |   9 | "node"    | "execution [done]"                 |      0 | "0"  | "()"  | true   |
```

later:

```
| 2016-02-07 19:21:48.800739821 +0000 +0000 | "29.506µs"    |   5 | "node"    | "raft"                             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:21:48.817891074 +0000 +0000 | "17.151253ms" |   6 | "node"    | "applying batch"                   |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:21:48.81789327 +0000 +0000  | "2.196µs"     |   7 | "node"    | "check sequence cache"             |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:21:48.817912259 +0000 +0000 | "18.989µs"    |   8 | "node"    | "execution"                        |      0 | "0"  | "()"  | true   |
| 2016-02-07 19:21:48.833296436 +0000 +0000 | "15.384177ms" |   9 | "node"    | "execution [done]"                 |      0 | "0"  | "()"  | true   |
```

In other words, Raft processing and the execution get slower and slower. Maybe I/O is sucky. Maybe two different things go wrong. Should be easy enough to hunt this down on Monday with some more elbow grease.

**Comment by tschottdorf** [comment]
Pushed that here for everyone else to try: https://github.com/cockroachdb/cockroach/tree/block_slugger

**Comment by mrtracy** [comment]
Alright, with Tobias's tracing I was able to narrow down the culprit.

tldr: introduced by baaba75bd33f0202ef61a64ef6c4e642807fd2bc

### Trace to narrow it down
These traces are very hacky and not checked in, so just know that they are at strategic locations.

```

 2016-02-09 03:22:06.438149717+00:00 | 92.947µs    |   9 | node      | execute                                                                                                |      0 | 0   | ()    | t
 2016-02-09 03:22:06.43815337+00:00  | 3.653µs     |  10 | node      | executing individual command *roachpb.BeginTransactionRequest                                          |      0 | 0   | ()    | t
 2016-02-09 03:22:06.438156499+00:00 | 3.129µs     |  11 | node      | BeginTransaction starts                                                                                |      0 | 0   | ()    | t
 2016-02-09 03:22:06.438158587+00:00 | 2.088µs     |  12 | node      | Transaction verified.                                                                                  |      0 | 0   | ()    | t
 2016-02-09 03:22:06.442732193+00:00 | 4.573606ms  |  13 | node      | Transaction was not present.                                                                           |      0 | 0   | ()    | t
 2016-02-09 03:22:06.442741914+00:00 | 9.721µs     |  14 | node      | checksum initialized.                                                                                  |      0 | 0   | ()    | t
 2016-02-09 03:22:06.442742331+00:00 | 417ns       |  15 | node      | Got put buffer.                                                                                        |      0 | 0   | ()    | t
 2016-02-09 03:22:06.442749832+00:00 | 7.501µs     |  16 | node      | mvccGetMetadata begin                                                                                  |      0 | 0   | ()    | t
 2016-02-09 03:22:06.44559688+00:00  | 2.847048ms  |  17 | node      | mvccGetMetadata seek complete                                                                          |      0 | 0   | ()    | t
 2016-02-09 03:22:06.44559754+00:00  | 660ns       |  18 | node      | getMetadata complete                                                                                   |      0 | 0   | ()    | t
 2016-02-09 03:22:06.445609668+00:00 | 12.128µs    |  19 | node      | Put internal complete.                                                                                 |      0 | 0   | ()    | t
 2016-02-09 03:22:06.44561002+00:00  | 352ns       |  20 | node      | Wrote transaction.                                                                                     |      0 | 0   | ()    | t
 2016-02-09 03:22:06.445612227+00:00 | 2.207µs     |  21 | node      | command complete                                                                                       |      0 | 0   | ()    | t
 2016-02-09 03:22:06.445618129+00:00 | 5.902µs     |  22 | node      | executing individual command *roachpb.ConditionalPutRequest                                            |      0 | 0   | ()    | t
 2016-02-09 03:22:06.445648646+00:00 | 30.517µs    |  23 | node      | Got put buffer.                                                                                        |      0 | 0   | ()    | t
```
All of the extra processing time is incurred during a call to `MVCCGet()` and a subsequent call to `MVCCPut`.  I was able to narrow down the MVCCPut() further - all of the extra time is taken in `mvccGetMetadata()`, specifically during a call to `iter.Seek()`.  I am certain that all of the extra time in the `MVCCGet()` is also consumed by the same `Seek()` method.

This pointed the finger at our rocksdb cgo layer, `db.cc`. This information made it *much* easier to perform a proper git bisect, which let me find the exact commit which introduced this performance problem:  baaba75bd33f0202ef61a64ef6c4e642807fd2bc 

That commit modifies cockroach's MVCCGet method to always use an iterator; the commit message indicates this was in an attempt to take advantage of some sort of cache used by the iterator in order to improve the speed of scans. However, it also introduces the progressive performance bug detailed in this issue.

I am not an expert in the behavior of RocksDB; it might be more productive for someone familiar with the system to investigate further.

**Comment by tschottdorf** [comment]
Good job Matt! @petermattis, all yours. We should put that behavior into a
benchmark as well.

**Comment by petermattis** [comment]
Lovely. Sounds like something is messed up with prefix iteration. Just to
confirm: the perf resets to good behavior if you switch to a new table?

That commit is 3 months old. I'm surprised you were able to git bisect that
far back.

**Comment by vivekmenezes** [comment]
Nice work Matt Tracy. Wanna write a bog post about this when it's fixed?

**Comment by petermattis** [comment]
My initial attempt at reproducing this with a test in the `storage/engine` package failed. My test is replicating the key and value structure of the `blocks` table and using the same logic to generate the `block_id`, `writer_id` and `block_num` fields. The test transactionally writes both the sentinel key and column key in a batch using `MVCCPut` and then resolves the intents in a second batch using `MVCCResolveWriteIntent`. Performance starts at 12k/sec and eventually levels off between 8k-10k/sec. I left the test running for 10min (it was killed by the default 10min test timeout) after which time it had written 3.9G of data.

So the bad news is the problem wasn't easily reproducible in a more focused setting. The good news is that there isn't some obvious brokenness with our current usage of RocksDB. Time to try to reproduce using the block_writer example.

**Comment by tschottdorf** [comment]
@petermattis maybe you're on the wrong track? I think the slow call is `BeginTransaction`, so look into a test that writes into the transaction table a lot? Just calling `BeginTransaction` with new transactions might be enough to repro.

**Comment by petermattis** [comment]
Hmm, a benchmark of the MVCC calls performed by `BeginTransaction` does not suffer serious slow down either. I am able to reproduce the performance degradation almost immediately using `block_writer`, and can see that `BeginTransaction` is taking milliseconds to complete. Still investigating.

**Comment by petermattis** [comment]
Somehow the bloom filters are not being used when running `block_writer`, but they are in the benchmark I wrote. In the benchmark I see a bloom filter utility rate of 98.8%. When running `block_writer` I never see the bloom filters accessed at all. (Rocksdb statistics can be enabled with `--vmodule=rocksdb=3`).

**Comment by petermattis** [comment]
The `block_writer` test isn't seeing any bloom filter usage because the data is still all in the memtable when performance starts to degraded. Instrumenting the code, I can see a single memtable seek taking >5ms. That shouldn't happen. The memtable is implemented as a skiplist. Lookups should be very fast.

**Comment by petermattis** [comment]
This is a little teaser. I'll write up what the problem was later (notice the timestamps below show the `block_writer` test has been running for ~4min):

```
2016/02/10 17:39:49 2956 dumps were executed at 2954.5/second (0 total errors)
...
2016/02/10 17:43:22 2365 dumps were executed at 2364.5/second (0 total errors)
```

**Comment by petermattis** [comment]
This was a fun little performance bug. TIL: Seeking into the RocksDB memtable can be very expensive in the presence of a large number of deletion tombstones.

Let's review what the test is doing. At its core, `block_writer` is writing "random" 256-1024 byte blocks into a table. Random here means that the primary key for the row in the table (an integer) is chosen randomly. Performance degrades quite rapidly which suggested some sort of horrible scaling problem as a table gets large. A little elbow grease and experimentation with the new `EXPLAIN (TRACE)` functionality @tschottdorf is working on pointed a finger not at the table data, but at writing the transaction record as part of the "begin transaction" operation that is being performed for every block written.

Additional instrumentation revealed that simply seeking to the key for the transaction record to see if it was present (which it wasn't because this was a brand new transaction) would sometimes take >20ms. This was very confusing. Checking to see if a key is present in RocksDB should be quick. 

My first step was to try and replicate the behavior of the "begin transaction" operation performed by `block_writer` in a smaller benchmark. I failed to find any performance anomalies. Curious.

Next up I narrowed down where the time was coming from. @mrtracy had already narrowed the culprit to the seek performed within [`mvccGetMetadata`](https://github.com/cockroachdb/cockroach/blob/master/storage/engine/mvcc.go#L593). I verified that this seek was showing horrible performance periodically when using the `block_writer` example. And the performance degraded over time, from merely bad to truly horrific.

So far the timing instrumentation had purely been performed at the Go level. Perhaps there was some performance issue with Cgo. Easy enough to test, we just need to push the timing instrumentation into our C++ code. So I instrumented [`DBIterSeek`](https://github.com/cockroachdb/cockroach/blob/master/storage/engine/rocksdb/db.cc#L1437) and saw exactly the same horrific performance that the seek in `mvccGetMetadata` is exhibiting. This was somewhat reassuring. Cgo was not the culprit.

So this was pointing to all of the time being spent inside of RocksDB. The `block_writer` test is writing a significant amount of data, that is its reason for existence. Perhaps that data was creating an excessive number of sstables in RocksDB which were causing the poor performance. Nope. In fact, there were no sstables when performance started to degrade, only the memtable! This was so surprising that I verified it twice, instrumenting RocksDB code to see that no memtable flushes had occurred and that only a memtable iterator was present during the `Seek`. 

Ok, wtf? We're inserting a bunch of keys into a memtable and the performance is steadily degrading for "seeking" to a key. Internally, the RocksDB memtable defaults to a [skip list](https://en.wikipedia.org/wiki/Skip_list). The skip list is neat data structure which is similar to a b-tree but can be implemented in a lock-free manner. The skip list in RocksDB requires synchronization on writes, but is safe for concurrent readers. When you see "skip list" you should be thinking fast-ordered-data-structure. How can the seeks within the skip list be taking milliseconds? A quick bit of instrumentation in [`SkipListRep::Iterator::Seek`](https://github.com/facebook/rocksdb/blob/08a78b6b44d7b699cbb0bc4c29516bd3f68f2ae9/memtable/skiplistrep.cc#L112) absolved it of blame. Individual seeks within the skip list were taking a handful of microseconds which was in line with my expectations.

So where are we at? We have only a single memtable containing all of the data. An iterator into this memtable is sometimes taking milliseconds to perform a seek operation. The memtable is implemented on top of a skip list and seeking inside the skip list is taking microseconds. The memtable and skip list are both concurrent for readers and no locks are being taken. So where is the time going?

This is actually the easy part of performance debugging. We have a reproducible test case and we have timing instrumentation which shows the problem at one level of the code and not at another. Ideally we'd perform a binary search of the code in between, narrowing our timing instrumentation until we identified by the culprit. Unfortunately, we have no easy tools to perform such a search so we have to do it manually. A little sleuthing reveals that [`DBIter::Seek`](https://github.com/facebook/rocksdb/blob/master/db/db_iter.cc#L707-L735) is the method which implements `rocksdb::Iterator::Seek`. A quick bit of timing instrumentation verifies this method exhibits the poor performance. RocksDB has many layers and `DBIter` is no exception. The bulk of `DBIter::Seek` is delegated to an internal `iter_` field. Another round of instrumentation reveals the `DBIter::iter_->Seek` call to be very fast (this is the call to `SkipListRep::Iterator::Seek`). Ah-ha! So the slowness is coming from elsewhere in `DBIter::Seek`.

`DBIter::FindNextUserEntry`, why you take so long? And what are you doing? RocksDB is a log-structured merge-tree. Unlike a b-tree where mutations are performed "in-place", a [log-structured merge-tree](https://en.wikipedia.org/wiki/Log-structured_merge-tree) performs edits similar to the way an accountant maintains finances: by adding records to indicate changes to the past instead of rewriting history. In RocksDB, when you overwrite a key you don't actually delete the old key but instead add a record for the new key that has a newer timestamp (sequence number) than the old key. When you delete a key you add a record (a deletion tombstone) that shadows the existence of the old key. 

`DBIter::FindNextUserEntry` is the top-most loop where processing of edits occurs. When `rocksdb::Iterator::Seek` is called, the resulting entry might point to a deleted key. `DBIter::FindNextUserEntry` is tasked with stepping over any deletions until it finds the next non-deleted entry.

This is what I know now. At the time I was debugging this I was trying to determine how `DBIter::FindNextUserEntry` could take milliseconds per call. That method is a big for-loop. How many iterations are performed? When performance is bad, 20k+. Whoa! Ok. That explains the bad performance. But why are there so many iterations? Deletion tombstones!

Here's the scenario: `block_writer` is writing random rows in a table. The performance of writing the transaction record associated with each row write degrades rapidly. Transaction records are written at a key derived from the first key written within the transaction and that key is deleted when the transaction is committed. So during random insertion into a table we have a segregated part of the key space where we're randomly inserting keys and almost immediately deleting them creating a large expanse of deletion tombstones. Before inserting the transaction key we're seeking within this deletion tombstone expanse and the seek needs to iterate from wherever it started to the next non-deleted key.

Imagine it this way: we have an array of elements. We insert elements into this array with a value of 1 and then we immediately mark them as deleted by flipping the element to 0. The result is an array composed entirely of zeroes. Every time we go to insert an element we first seek to where that element will be inserted into the array and then iterate forward until we find the next non-zero element. It should be fairly clear that performance will quickly be dominated by that iteration.

So seeking into a memtable with a large contiguous region of deleted keys is a bad idea. What can we do to fix the problem? It turns out that RocksDB has a solution readily at hand: when you create a `rocksdb::Iterator` you can specify an upper-bound past which the iterator will not iterate (see [`rocksdb::ReadOptions::iterate_upper_bound`](https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L1357)). CockroachDB's MVCC layer was already taking advantage of RocksDB prefix iteration (which is a topic for another time), but the short story is that we're already indicating in the code where we're going to create an iterator that has a restricted range for iteration. It was straightforward to modify this code to allow specification of `rocksdb::ReadOptions::iterate_upper_bound` when the RocksDB iterator was created.

Prior to this fix we see the following performance from `block_writer`:

```
2016/02/10 22:10:33 1983 dumps were executed at 1978.2/second (0 total errors)
2016/02/10 22:10:34 1067 dumps were executed at 1064.5/second (0 total errors)
2016/02/10 22:10:35 798 dumps were executed at 798.1/second (0 total errors)
2016/02/10 22:10:36 633 dumps were executed at 633.0/second (0 total errors)
2016/02/10 22:10:37 561 dumps were executed at 563.4/second (0 total errors)
2016/02/10 22:10:38 452 dumps were executed at 449.9/second (0 total errors)
2016/02/10 22:10:39 420 dumps were executed at 421.6/second (0 total errors)
2016/02/10 22:10:40 405 dumps were executed at 403.5/second (0 total errors)
2016/02/10 22:10:41 362 dumps were executed at 362.1/second (0 total errors)
2016/02/10 22:10:42 351 dumps were executed at 352.2/second (0 total errors)
```

After the fix we have:

```
2016/02/10 22:12:17 2860 dumps were executed at 2858.7/second (0 total errors)
2016/02/10 22:12:18 2881 dumps were executed at 2881.1/second (0 total errors)
2016/02/10 22:12:19 2686 dumps were executed at 2685.9/second (0 total errors)
2016/02/10 22:12:20 2811 dumps were executed at 2801.6/second (0 total errors)
2016/02/10 22:12:21 2808 dumps were executed at 2817.7/second (0 total errors)
2016/02/10 22:12:22 2691 dumps were executed at 2690.7/second (0 total errors)
2016/02/10 22:12:23 2786 dumps were executed at 2786.3/second (0 total errors)
2016/02/10 22:12:24 2791 dumps were executed at 2791.0/second (0 total errors)
2016/02/10 22:12:25 2793 dumps were executed at 2792.9/second (0 total errors)
2016/02/10 22:12:26 2815 dumps were executed at 2815.1/second (0 total errors)
```

**Comment by tamird** [comment]
Nice writeup!

> when you create a rocksdb::Iterator you can specify an upper-bound past which the iterator will not iterate

Does this mean we stop iterating early in `DBIter::FindNextUserEntry`, and write the new transaction record on top of an old tombstone?

**Comment by vivekmenezes** [comment]
Thanks for sharing!

**Comment by bdarnell** [comment]
Nice detective work!

Even with the restricted iterator, these 20k+ tombstones will get written to disk when a minor compaction happens (and potentially rewritten in future compactions until the system can determine they are completely obsolete). I believe that since we use bloom filters, we have a solution for this: set the [`filter_deletes`](https://github.com/facebook/rocksdb/blob/908100399cdc8855980ca9f8c05d37836857ab79/include/rocksdb/options.h#L595-L603) option. This means that on delete, rocksdb will consult the bloom filter first and will only create tombstones for keys that the filter indicates may exist on disk. Keys that are created and deleted in between minor compactions can be deleted from the memtable without leaving any trace on disk. (Bigtable had the same optimization, but there wasn't a separate option for it: it was enabled whenever bloom filters were enabled)

**Comment by tschottdorf** [comment]
Good job, Pete.
If these are 1PC txns, we could also omit writing the txn record altogether.

**Comment by tschottdorf** [comment]
> Does this mean we stop iterating early in DBIter::FindNextUserEntry, and write the new transaction record on top of an old tombstone?

I think the slow `Seek` is just us looking for a meta entry. Before the change, what we get back from the underlying iterator is the next existing key `>=` the encoded metadata key. With the change, the iterator will simply be invalid after `Seek`, which basically tells us there is no meta entry for our key - all we needed to know (we care only about the `==` in `>=`).

**Comment by petermattis** [comment]
@tamird what @tschottdorf said is correct.

@tschottdorf Yes, these are 1PC txns. Not writing the txn record at all seems like a good optimization. Can we also avoid writing the intents? Is that done already? Since we're doing all of the writes in a single batch that seems safe as well.

@bdarnell Yeah, we should take a look at the `filter_deletes` option. I also discovered that there is support for bloom filters in the memtable. Not sure if this could benefit us or not, but worth exploring.

Lastly, I [discovered](https://github.com/facebook/rocksdb/issues/983) and [fixed](https://github.com/facebook/rocksdb/pull/984) a tiny bug in RocksDB.

**Comment by tschottdorf** [comment]
Filed #4324 for the 1PC optimization.

**Comment by igorcanadi** [comment]
> TIL: Seeking into the RocksDB memtable can be very expensive in the presence of a large number of deletion tombstones.

Tombstones are the biggest problem when dealing with write-optimized data stores. :( They can cause high P99 latencies, especially when you're using LSM to implement a queue (get-first-element of a queue when bunch of first elements are deleted...).

We have some optimizations in place to help with that. I see you've already found one (iterate_upper_bound). There is another optimization that we call SingleDelete(), and that boosted our MySQL performance (under our workload). Check out a blog post about this: http://smalldatum.blogspot.com/2016/01/rocksdb-vs-innodb-via-linkbench.html

The way we debug tombstone-related problems is by using PerfContext: https://github.com/facebook/rocksdb/blob/master/include/rocksdb/perf_context.h (count of tombstones is a cryptic name `internal_delete_skipped_count`). It's a neat way of tracking what a single transaction is doing. If you pipe the data to some kind of columnar store, you have a very powerful performance analysis story.

**Comment by petermattis** [comment]
I saw the `PerfContext` stuff, but unfortunately I wasn't able to access it from our C++ due to some weird linker error about a thread-local variable. This might be Mac OS X or Go specific, I need to investigate further.

Thanks for the pointer to `SingleDelete`. I'll have to read more to understand if we can utilize it or not.
